erigon-lib/diagnostics/headers.go

```diff
@@ -0,0 +1,115 @@
+package diagnostics
+
+import (
+	"context"
+
+	"github.com/ledgerwatch/erigon-lib/common"
+	"github.com/ledgerwatch/log/v3"
+)
+
+func (d *DiagnosticClient) setupHeadersDiagnostics() {
+	d.runHeadersWaitingListener()
+	d.runWriteHeadersListener()
+	d.runCanonicalMarkerListener()
+	d.runProcessedListener()
+}
+
+func (d *DiagnosticClient) GetHeaders() Headers {
+	return d.headers
+}
+
+func (d *DiagnosticClient) runHeadersWaitingListener() {
+	go func() {
+		ctx, ch, cancel := Context[HeadersWaitingUpdate](context.Background(), 1)
+		defer cancel()
+
+		rootCtx, _ := common.RootContext()
+
+		StartProviders(ctx, TypeOf(HeadersWaitingUpdate{}), log.Root())
+		for {
+			select {
+			case <-rootCtx.Done():
+				cancel()
+				return
+			case info := <-ch:
+				d.headerMutex.Lock()
+				d.headers.WaitingForHeaders = info.From
+				d.headerMutex.Unlock()
+
+				return
+			}
+		}
+	}()
+}
+
+func (d *DiagnosticClient) runWriteHeadersListener() {
+	go func() {
+		ctx, ch, cancel := Context[BlockHeadersUpdate](context.Background(), 1)
+		defer cancel()
+
+		rootCtx, _ := common.RootContext()
+
+		StartProviders(ctx, TypeOf(BlockHeadersUpdate{}), log.Root())
+		for {
+			select {
+			case <-rootCtx.Done():
+				cancel()
+				return
+			case info := <-ch:
+				d.headerMutex.Lock()
+				d.headers.WriteHeaders = info
+				d.headerMutex.Unlock()
+
+				return
+			}
+		}
+	}()
+}
+
+func (d *DiagnosticClient) runCanonicalMarkerListener() {
+	go func() {
+		ctx, ch, cancel := Context[HeaderCanonicalMarkerUpdate](context.Background(), 1)
+		defer cancel()
+
+		rootCtx, _ := common.RootContext()
```

Comment:
1. `common.RootContext()` - it's not global variable access - it's constructor. and better don't have more than 1 root context. 
please pass existing root context to `DiagnosticClient` constructor. save it as field and use here. 
(feel free to rename RootContext to NewRootContext)


2. cancel 1 context based on another one - it's anti-pattern:  
```
case <-rootCtx.Done():
				cancel()
```
better establish child-parent relationship between contexts: `Context[HeaderCanonicalMarkerUpdate](context.Background()` -> `Context[HeaderCanonicalMarkerUpdate](rootCtx`

Reply:
done